python/lsst/daf/butler/tests/_testRepo.py

```diff
@@ -311,14 +313,88 @@ def expandUniqueId(butler: Butler, partialId: Mapping[str, Any]) -> DataCoordina
+        already existed.
+    """
+    # No values given by caller for this dimension.  See if any exist
+    # in the registry that are consistent with the values of dimensions
+    # we do have:
+    match_data_id = {key: data_id[key] for key in data_id.keys() & dimension.dimensions.names}
+    matches = list(butler.registry.queryDimensionRecords(dimension, dataId=match_data_id).limit(1))
+    if not matches:
+        # Nothing in the registry matches: invent a data ID value
+        # with the right type (actual value does not matter).
+        # We may or may not actually make a record with this; that's
+        # easier to check later.
+        dimension_value = _makeRandomDataIdValue(dimension)
+        return dimension_value, True
+    else:
+        # A record does exist in the registry.  Use its data ID value,
+        # and skip the rest of the record-fabrication logic.
+        return matches[0].dataId[dimension.name], False
+
+
+def _makeDimensionRecordDict(data_id: dict[str, Union[str, int]], dimension: Dimension) -> dict[str, Any]:
+    """Create a dictionary that can be used to build a `DimensionRecord` that
+    is consistent with the given data ID.
+
+    Parameters
+    ----------
+    data_id : `dict` [ `str`, `str` or `int` ]
+        Dictionary that contains values for all of
+        ``dimension.dimensions.names`` (the main dimension, its recursive
+        required dependencies, and its non-recursive implied dependencies).
+    dimension : `Dimension`
+        Dimension to build a record dictionary for.
+
+    Returns
+    -------
+    record_dict : `dict` [ `str`, `object` ]
+        Dictionary that can be passed as ``**kwargs`` to this dimensions
+        record class constructor.
+    """
+    # Add the primary key field for this dimension.
+    record_dict: dict[str, Any] = {dimension.primaryKey.name: data_id[dimension.name]}
+    # Define secondary keys (e.g., detector name given detector id)
+    record_dict.update(_fillAllKeys(dimension, data_id[dimension.name]))
+    # Set the foreign key values for any related dimensions that should
+    # appear in the record.
+    for related_dimension in dimension.dimensions:
+        if related_dimension.name != dimension.name:
+            record_dict[related_dimension.name] = data_id[related_dimension.name]
+    return record_dict
+
+
 def addDataIdValue(butler: Butler, dimension: str, value: Union[str, int], **related: Union[str, int]):
     """Add a new data ID to a repository.
 
     Related dimensions (e.g., the instrument associated with a detector) may be
     specified using ``related``. While these keywords are sometimes needed to
     get self-consistent repositories, you do not need to define relationships
-    you do not use. Any unspecified dimensions will be linked arbitrarily,
-    withe values created randomly when needed.
+    you do not use.  Any unspecified dimensions implied will be linked
```

Comment:
I've rewritten this block and merged it into to the `Notes` section, as it was getting pretty long.  Here's the new complete `Notes` text (the second paragraph is unchanged):

> Related dimensions (e.g., the instrument associated with a detector) may be
    specified using ``related``, which requires a value for those dimensions to
    have been added to the repository already (generally with a previous call
    to `addDataIdValue`.  Any dependencies of the given dimension that are not
    included in ``related`` will be linked to existing values arbitrarily, and
    (for implied dependencies only) created and also inserted into the registry
    if they do not exist.  Values for required dimensions and those given in
    ``related`` are never created.
>
>    Because this function creates filler data, it is only suitable for test
    repositories. It should not be used for repositories intended for real data
    processing or analysis, which have known dimension values.